src/steamship/agents/tools/question_answering/vector_search_qa_tool.py

```diff
@@ -43,17 +44,31 @@ def answer_question(self, question: str, context: AgentContext) -> List[Block]:
         task.wait()
 
         source_texts = []
+        source_metadata = []
 
         for item in task.output.items:
             if item.tag and item.tag.text:
                 item_data = {"text": item.tag.text}
                 source_texts.append(self.source_document_prompt.format(**item_data))
+                _metadata = {}
+                if item.tag.value:
+                    _metadata.update(item.tag.value)
+                source_metadata.append(_metadata)
 
         final_prompt = self.question_answering_prompt.format(
             **{"source_text": "\n".join(source_texts), "question": question}
         )
 
-        return get_llm(context, default=OpenAI(client=context.client)).complete(prompt=final_prompt)
+        output_blocks = get_llm(context, default=OpenAI(client=context.client)).complete(
+            prompt=final_prompt
+        )
+        for output_block in output_blocks:
```

Comment:
Looking at this again, i wonder if this isn't slightly cheating / too clever -- The LLM isn't telling us which sources it used, right?  So, we are just saying which sources we found and added to the query. But what if the LLM ignored the sources and/or only used certain ones?

Reply:
I'm almost positive I've seen some prompts ask the LLM to use citations for exactly this reason, though I don't know how well it works -- I can look into that and do a few test runs to see how stable the output is.